File: ruby-sdk/lib/cleanster/api/bookings_api.rb
module Cleanster
  module Api
    # BookingsApi — manage the full lifecycle of cleaning appointments.
    class BookingsApi
      def initialize(http)
        @http = http
      end

      # List upcoming and past bookings.
      #
      # @param page_no [Integer, nil] Page number (1-based). Pass nil for page 1.
      # @param status  [String, nil]  Filter: OPEN | CLEANER_ASSIGNED | COMPLETED | CANCELLED | REMOVED
      # @return [Models::ApiResponse]
      def get_bookings(page_no: nil, status: nil)
        raw = @http.get("/v1/bookings", params: { pageNo: page_no, status: status })
        Models::ApiResponse.from_hash(raw)
      end

      # Schedule a new cleaning appointment.
      #
      # @param request [Hash] date, time, property_id, room_count, bathroom_count,
      #                       plan_id, hours, extra_supplies, payment_method_id.
      # @return [Models::ApiResponse<Models::Booking>]
      def create_booking(request)
        raw = @http.post("/v1/bookings/create", body: camel_keys(request))
        Models::ApiResponse.from_hash(raw, model_class: Models::Booking)
      end

      # Get full details of a specific booking.
      #
      # @param booking_id [Integer]
      # @return [Models::ApiResponse<Models::Booking>]
      def get_booking_details(booking_id)
        raw = @http.get("/v1/bookings/#{booking_id}")
        Models::ApiResponse.from_hash(raw, model_class: Models::Booking)
      end

      # Cancel a booking.
      #
      # @param booking_id [Integer]
      # @param reason     [String, nil] Optional cancellation reason.
      # @return [Models::ApiResponse]
      def cancel_booking(booking_id, reason: nil)
        body = reason ? { reason: reason } : {}
        raw  = @http.post("/v1/bookings/#{booking_id}/cancel", body: body)
        Models::ApiResponse.from_hash(raw)
      end

      # Reschedule a booking to a new date and time.
      #
      # @param booking_id [Integer]
      # @param date [String] New date in YYYY-MM-DD format.
      # @param time [String] New time in HH:mm format (24-hour).
      # @return [Models::ApiResponse]
      def reschedule_booking(booking_id, date:, time:)
        raw = @http.post("/v1/bookings/#{booking_id}/reschedule", body: { date: date, time: time })
        Models::ApiResponse.from_hash(raw)
      end

      # Manually assign a specific cleaner to a booking.
      #
      # @param booking_id [Integer]
      # @param cleaner_id [Integer]
      # @return [Models::ApiResponse]
      def assign_cleaner(booking_id, cleaner_id:)
        raw = @http.post("/v1/bookings/#{booking_id}/cleaner", body: { cleanerId: cleaner_id })
        Models::ApiResponse.from_hash(raw)
      end

      # Remove the currently assigned cleaner from a booking.
      #
      # @param booking_id [Integer]
      # @return [Models::ApiResponse]
      def remove_assigned_cleaner(booking_id)
        raw = @http.delete("/v1/bookings/#{booking_id}/cleaner")
        Models::ApiResponse.from_hash(raw)
      end

      # Adjust the number of hours for a booking.
      #
      # @param booking_id [Integer]
      # @param hours [Numeric]
      # @return [Models::ApiResponse]
      def adjust_hours(booking_id, hours:)
        raw = @http.post("/v1/bookings/#{booking_id}/hours", body: { hours: hours })
        Models::ApiResponse.from_hash(raw)
      end

      # Pay outstanding expenses for a completed booking (within 72h of completion).
      #
      # @param booking_id        [Integer]
      # @param payment_method_id [Integer]
      # @return [Models::ApiResponse]
      def pay_expenses(booking_id, payment_method_id:)
        raw = @http.post("/v1/bookings/#{booking_id}/expenses", body: { paymentMethodId: payment_method_id })
        Models::ApiResponse.from_hash(raw)
      end

      # Get the inspection report for a completed booking.
      def get_booking_inspection(booking_id)
        raw = @http.get("/v1/bookings/#{booking_id}/inspection")
        Models::ApiResponse.from_hash(raw)
      end

      # Get detailed inspection information for a completed booking.
      def get_booking_inspection_details(booking_id)
        raw = @http.get("/v1/bookings/#{booking_id}/inspection/details")
        Models::ApiResponse.from_hash(raw)
      end

      # Attach a cleaning checklist to a specific booking.
      #
      # @param booking_id   [Integer]
      # @param checklist_id [Integer]
      # @return [Models::ApiResponse]
      def assign_checklist_to_booking(booking_id, checklist_id)
        raw = @http.put("/v1/bookings/#{booking_id}/checklist/#{checklist_id}")
        Models::ApiResponse.from_hash(raw)
      end

      # Submit a star rating and optional comment after a booking completes.
      #
      # @param booking_id [Integer]
      # @param rating     [Integer] 1–5
      # @param comment    [String, nil]
      # @return [Models::ApiResponse]
      def submit_feedback(booking_id, rating:, comment: nil)
        body = { rating: rating }
        body[:comment] = comment if comment
        raw = @http.post("/v1/bookings/#{booking_id}/feedback", body: body)
        Models::ApiResponse.from_hash(raw)
      end

      # Add a tip for the cleaner (within 72h of booking completion).
      #
      # @param booking_id        [Integer]
      # @param amount            [Numeric]
      # @param payment_method_id [Integer]
      # @return [Models::ApiResponse]
      def add_tip(booking_id, amount:, payment_method_id:)
        raw = @http.post("/v1/bookings/#{booking_id}/tip",
                         body: { amount: amount, paymentMethodId: payment_method_id })
        Models::ApiResponse.from_hash(raw)
      end

      # Retrieve all chat messages for a booking thread.
      def get_chat(booking_id)
        raw = @http.get("/v1/bookings/#{booking_id}/chat")
        Models::ApiResponse.from_hash(raw)
      end

      # Post a chat message in a booking thread.
      #
      # @param booking_id [Integer]
      # @param message    [String]
      # @return [Models::ApiResponse]
      def send_message(booking_id, message:)
        raw = @http.post("/v1/bookings/#{booking_id}/chat", body: { message: message })
        Models::ApiResponse.from_hash(raw)
      end

      # Delete a specific chat message.
      #
      # @param booking_id [Integer]
      # @param message_id [String]
      # @return [Models::ApiResponse]
      def delete_message(booking_id, message_id)
        raw = @http.delete("/v1/bookings/#{booking_id}/chat/#{message_id}")
        Models::ApiResponse.from_hash(raw)
      end

      private

      def camel_keys(hash)
        map = {
          property_id:        "propertyId",
          room_count:         "roomCount",
          bathroom_count:     "bathroomCount",
          plan_id:            "planId",
          extra_supplies:     "extraSupplies",
          payment_method_id:  "paymentMethodId",
          coupon_code:        "couponCode"
        }
        hash.transform_keys { |k| map[k] || k }
      end
    end
  end
end

File: php-sdk/src/Api/BookingsApi.php
<?php

declare(strict_types=1);

namespace Cleanster\Api;

use Cleanster\HttpClient;
use Cleanster\Models\ApiResponse;
use Cleanster\Models\Booking;

/**
 * Manages the full lifecycle of cleaning appointments.
 */
final class BookingsApi
{
    public function __construct(private readonly HttpClient $http) {}

    // -------------------------------------------------------------------------
    // Listing and retrieval
    // -------------------------------------------------------------------------

    /**
     * Retrieve a paginated list of bookings.
     *
     * @param int|null    $pageNo Optional page number (1-based).
     * @param string|null $status Optional status filter.
     *                            One of: OPEN | CLEANER_ASSIGNED | COMPLETED | CANCELLED | REMOVED
     *
     * @return ApiResponse<Booking[]>
     */
    public function getBookings(?int $pageNo = null, ?string $status = null): ApiResponse
    {
        $query = [];
        if ($pageNo !== null) {
            $query['pageNo'] = $pageNo;
        }
        if ($status !== null) {
            $query['status'] = $status;
        }
        $raw = $this->http->get('/v1/bookings', $query);
        return $this->wrapList($raw, Booking::class);
    }

    /**
     * Schedule a new cleaning appointment.
     *
     * @param array $request {
     *   @type string $date            Required — YYYY-MM-DD
     *   @type string $time            Required — HH:mm (24-hour)
     *   @type int    $propertyId      Required
     *   @type int    $roomCount       Required
     *   @type int    $bathroomCount   Required
     *   @type int    $planId          Required — from getPlans()
     *   @type float  $hours           Required — from getRecommendedHours()
     *   @type bool   $extraSupplies   Required
     *   @type int    $paymentMethodId Required
     *   @type string $couponCode      Optional
     *   @type int[]  $extras          Optional
     * }
     * @return ApiResponse<Booking>
     */
    public function createBooking(array $request): ApiResponse
    {
        $raw = $this->http->post('/v1/bookings/create', $request);
        return $this->wrapModel($raw, Booking::class);
    }

    /**
     * Return full details of a specific booking.
     *
     * @return ApiResponse<Booking>
     */
    public function getBookingDetails(int $bookingId): ApiResponse
    {
        $raw = $this->http->get("/v1/bookings/{$bookingId}");
        return $this->wrapModel($raw, Booking::class);
    }

    // -------------------------------------------------------------------------
    // Lifecycle mutations
    // -------------------------------------------------------------------------

    /**
     * Cancel a booking.
     *
     * @param int         $bookingId Booking to cancel.
     * @param string|null $reason    Optional cancellation reason.
     */
    public function cancelBooking(int $bookingId, ?string $reason = null): ApiResponse
    {
        $body = [];
        if ($reason !== null && $reason !== '') {
            $body['reason'] = $reason;
        }
        $raw = $this->http->post("/v1/bookings/{$bookingId}/cancel", $body ?: null);
        return $this->wrapRaw($raw);
    }

    /**
     * Move a booking to a new date and time.
     *
     * @param int    $bookingId Booking to reschedule.
     * @param string $date      New date (YYYY-MM-DD).
     * @param string $time      New time (HH:mm).
     */
    public function rescheduleBooking(int $bookingId, string $date, string $time): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/reschedule", [
            'date' => $date,
            'time' => $time,
        ]);
        return $this->wrapRaw($raw);
    }

    /** Manually assign a specific cleaner to a booking. */
    public function assignCleaner(int $bookingId, int $cleanerId): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/cleaner", ['cleanerId' => $cleanerId]);
        return $this->wrapRaw($raw);
    }

    /** Remove the currently assigned cleaner from a booking. */
    public function removeAssignedCleaner(int $bookingId): ApiResponse
    {
        $raw = $this->http->delete("/v1/bookings/{$bookingId}/cleaner");
        return $this->wrapRaw($raw);
    }

    /** Update the number of hours for a booking. */
    public function adjustHours(int $bookingId, float $hours): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/hours", ['hours' => $hours]);
        return $this->wrapRaw($raw);
    }

    // -------------------------------------------------------------------------
    // Post-completion actions
    // -------------------------------------------------------------------------

    /** Pay outstanding expenses within 72 hours of booking completion. */
    public function payExpenses(int $bookingId, int $paymentMethodId): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/expenses", [
            'paymentMethodId' => $paymentMethodId,
        ]);
        return $this->wrapRaw($raw);
    }

    /** Return the inspection report for a completed booking. */
    public function getBookingInspection(int $bookingId): ApiResponse
    {
        $raw = $this->http->get("/v1/bookings/{$bookingId}/inspection");
        return $this->wrapRaw($raw);
    }

    /** Return detailed inspection information for a completed booking. */
    public function getBookingInspectionDetails(int $bookingId): ApiResponse
    {
        $raw = $this->http->get("/v1/bookings/{$bookingId}/inspection/details");
        return $this->wrapRaw($raw);
    }

    /**
     * Attach a checklist to a specific booking, overriding the property's default.
     */
    public function assignChecklistToBooking(int $bookingId, int $checklistId): ApiResponse
    {
        $raw = $this->http->put("/v1/bookings/{$bookingId}/checklist/{$checklistId}");
        return $this->wrapRaw($raw);
    }

    /**
     * Submit a star rating (1–5) and optional comment after a booking completes.
     *
     * @param int         $bookingId Completed booking ID.
     * @param int         $rating    Star rating 1–5.
     * @param string|null $comment   Optional comment.
     */
    public function submitFeedback(int $bookingId, int $rating, ?string $comment = null): ApiResponse
    {
        $body = ['rating' => $rating];
        if ($comment !== null && $comment !== '') {
            $body['comment'] = $comment;
        }
        $raw = $this->http->post("/v1/bookings/{$bookingId}/feedback", $body);
        return $this->wrapRaw($raw);
    }

    /** Add a tip for the cleaner within 72 hours of booking completion. */
    public function addTip(int $bookingId, float $amount, int $paymentMethodId): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/tip", [
            'amount'          => $amount,
            'paymentMethodId' => $paymentMethodId,
        ]);
        return $this->wrapRaw($raw);
    }

    // -------------------------------------------------------------------------
    // Chat
    // -------------------------------------------------------------------------

    /** Retrieve all chat messages in a booking thread. */
    public function getChat(int $bookingId): ApiResponse
    {
        $raw = $this->http->get("/v1/bookings/{$bookingId}/chat");
        return $this->wrapRaw($raw);
    }

    /** Post a chat message to a booking thread. */
    public function sendMessage(int $bookingId, string $message): ApiResponse
    {
        $raw = $this->http->post("/v1/bookings/{$bookingId}/chat", ['message' => $message]);
        return $this->wrapRaw($raw);
    }

    /** Delete a specific chat message from a booking thread. */
    public function deleteMessage(int $bookingId, string $messageId): ApiResponse
    {
        $raw = $this->http->delete("/v1/bookings/{$bookingId}/chat/{$messageId}");
        return $this->wrapRaw($raw);
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private function wrapModel(array $raw, string $class): ApiResponse
    {
        return new ApiResponse(
            $raw['status'] ?? 200,
            $raw['message'] ?? 'OK',
            new $class($raw['data'] ?? [])
        );
    }

    private function wrapList(array $raw, string $class): ApiResponse
    {
        $items = array_map(fn(array $item) => new $class($item), $raw['data'] ?? []);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $items);
    }

    private function wrapRaw(array $raw): ApiResponse
    {
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }
}
